src/main/java/org/spongepowered/common/mixin/core/entity/item/MixinEntityArmorStand.java

```diff
@@ -108,4 +123,87 @@ public void supplyVanillaManipulators(List<DataManipulator<?, ?>> manipulators)
         manipulators.add(getArmorStandData());
     }
+
+    /**
+     * The return value is set to false if the entity should not be completely
+     * destroyed.
+     */
+    private void fireDestroyDamageEvent(DamageSource source, CallbackInfoReturnable<Boolean> cir) {
+        try (CauseStackManager.StackFrame frame = Sponge.getCauseStackManager().pushCauseFrame()) {
+            DamageEventHandler.generateCauseFor(source);
+            DamageEntityEvent event = SpongeEventFactory.createDamageEntityEvent(Sponge.getCauseStackManager().getCurrentCause(), new ArrayList<>(),
+                    this, Math.max(1000, this.getHealth()));
+            if (SpongeImpl.postEvent(event)) {
+                cir.setReturnValue(false);
+            }
+            if (event.getFinalDamage() < this.getHealth()) {
+                this.damageArmorStand((float) event.getFinalDamage());
+                cir.setReturnValue(false);
+            }
+        }
+    }
+
+    @Inject(method = "attackEntityFrom", at = @At(value = "INVOKE", target = "Lnet/minecraft/entity/item/EntityArmorStand;setDead()V", ordinal = 0),
+            cancellable = true)
+    private void fireDamageEventOutOfWorld(DamageSource source, float amount, CallbackInfoReturnable<Boolean> cir) {
+        fireDestroyDamageEvent(source, cir);
+    }
+
+    @Inject(method = "attackEntityFrom", at = @At(value = "INVOKE", target = "Lnet/minecraft/entity/item/EntityArmorStand;dropContents()V"),
+            cancellable = true)
+    private void fireDamageEventExplosion(DamageSource source, float amount, CallbackInfoReturnable<Boolean> cir) {
+        fireDestroyDamageEvent(source, cir);
+    }
+
+    @Redirect(method = "attackEntityFrom", at = @At(value = "INVOKE", target = "Lnet/minecraft/entity/item/EntityArmorStand;damageArmorStand(F)V"))
+    private void fireDamageEventDamage(EntityArmorStand self, float effectiveAmount, DamageSource source, float originalAmount) {
+        try (CauseStackManager.StackFrame frame = Sponge.getCauseStackManager().pushCauseFrame()) {
+            DamageEventHandler.generateCauseFor(source);
+            DamageEntityEvent event = SpongeEventFactory.createDamageEntityEvent(Sponge.getCauseStackManager().getCurrentCause(), new ArrayList<>(),
+                    this, effectiveAmount);
+            if (!SpongeImpl.postEvent(event)) {
+                this.damageArmorStand((float) event.getFinalDamage());
+            }
+        }
+    }
+
+    @Inject(method = "attackEntityFrom", at = @At(value = "INVOKE", target = "Lnet/minecraft/entity/item/EntityArmorStand;playBrokenSound()V"),
+            cancellable = true)
+    private void fireDamageEventCreativePunch(DamageSource source, float amount, CallbackInfoReturnable<Boolean> cir) {
+        fireDestroyDamageEvent(source, cir);
+    }
+
+    @Inject(method = "attackEntityFrom", at = @At(value = "INVOKE", target = "Lnet/minecraft/world/World;setEntityState(Lnet/minecraft/entity/Entity;B)V"),
```

Comment:
This injection and the next injection appear to inject at the same point, maybe use a slice if they aren't? Or make it clearer where it's injecting into?

All of the injections themselves seem out of place if we can't just overwrite the method (though that also has drawbacks). I'm somewhat keen on an idea that we should start using slices more often to define where a method we're expecting to inject into should exist.

Reply:
Whoops. I meant `dropBlock()` instead.

I also threw in slices for a bunch of the methods.